web/app.js

```diff
@@ -2218,6 +2253,14 @@ function webViewerKeyDown(evt) {
     switch (evt.keyCode) {
       case 38: // up arrow
       case 33: // pg up
+        if (!pdfViewer.isVerticalScrollbarEnabled) {
+          if (PDFViewerApplication.page > 1) {
+            PDFViewerApplication.page--;
+          }
+          handled = true;
+          break;
+        }
+        /* falls through */
```

Comment:
This changes the default behavior without enabling any of your new options. Could you explain why you made those changes to the default, and why it's okay to change the behavior?

Reply:
This was a response to [this request](/mozilla/pdf.js/pull/9208#issuecomment-362788050). It seemed fair that, just like left and right mean previous and next page when there's no horizontal scroll bar (because otherwise, when there's no horizontal scroll bar, left and right can't do anything, so why would the user press them unless their intent was to turn pages?), that up/down should mean previous/next page when there's no vertical scroll bar. (Otherwise, the only way to turn pages from the keyboard when in horizontal scrolling mode is to use the letter keys, which aren't as discoverable to most users as the arrow/page-navigation keys, IMO.)

The only way this ‘changes’ behavior without enabling any new options is when the entire PDF fits vertically in the viewport (so there's no vertical scrollbar), or when in presentation mode. If the PDF fits vertically, then the effect of going to the previous page, while visible (in the current page number box), is harmless IMO. If in presentation mode, there actually is no change; the instruction pointer would have eventually fallen through to turning the page anyway.

Reply:
Thanks for this additional explanation. My eyes were drawn to this code exactly because of the repetition of the tail of the fall-through-cases.

Since your code only affects presentation mode, in the sense that arrow-up/PgUp does the expected thing, I think that you can merge the condition with `if (pdfViewer.isHorizontalScrollbarEnabled) {` in the `case 37: // left arrow` block. The effect is equivalent, but the code is simpler.

Reply:
I'm sorry, I don't think I understand. Inside `case 37:`, we don't know if we've fallen through from an arrow-up or page-up, or if the event was actually an arrow-left, so `evt.keyCode` would need to be tested again in order to know if we should check for `isHorizontalScrollbarEnabled` or `isVerticalScrollbarEnabled`. Is that what you're recommending?

Reply:
My reasoning is as follows:

- The new logic is only relevant when there is no vertical scrollbar. In a normal view, without a scrollbar, there is only one page - not really interesting. So the remaining point of interest is when presentation mode is enabled.
- The early break in the backspace case can be ignored sincec it applues to non-presentation mode.
- case 75 / 80 is the terminal block, and the only block in between backspace and the terminal block is a check plus an optional early break. The body (logic) of the terminal block is identical to the body of the if-block that you added. So, you can merge your condition with the if-condition in the penultimate block, and end up with `if (pdfViewer.isHorizontalScrollbarEnabled || pdfViewer.isVerticalScrollbarEnabled) {` (and another comment).

Reply:
If I did that, wouldn't left arrow stop working as previous page in vertical scrolling mode? We'd go to case 37, see that there is a vertical scrollbar, and break.

Reply:
That's right, I had not accounted for the new vertical scrolling mode yet. After thinking a bit more about it, there are no scrollbars at all in presentation mode, so really the only case to cover is your new horizontal scroling feature.

That can be accounted for by modifying the backspace case to the following (and have a `scrollMode` getter on `PDFViewer`):

```javascript
      case 8: // backspace
        if (!isViewerInPresentationMode) {
          if (pdfViewer.scrollMode === ScrollMode.VERTICAL) {
            if (pdfViewer.currentScaleValue !== 'page-fit') {
              break;
            }
          } else {
            // vertical scrolling with keys in horizontal scrolling mode
            if (isVerticalScrollbarEnabled) {
              break;
            }
          }
        }
        /* falls through */
      case 37: // left arrow
```

This has a few more lines, but captures the intended logic more clearly without code duplication.

There is another thing: In page fit mode, even if there is somehow an (unexpected) horizontal scrollbar (with default (vertical) scrolling mode), the arrow up/down keys are jumping to the previous/next page. If you want to have a similar exception for horizontal scrolling mode, then you need to adjust the above clauses for that.

When I test your current demo, the page-fit mode does not actually fit in Chromium 64 and Firefox 58 on Linux: There is still a visible vertical scrollbar.

Reply:
That still wouldn't work without more changes; in horizontal scrolling mode, we'd fall through to case 37, which checks if the horizontal scroll bar is enabled and then breaks, leading to up arrow not causing a page turn. Besides, I strongly disagree that this captures the intended logic clearly. The scroll mode is not directly relevant; what is relevant is whether or not there are vertical/horizontal scroll bars. You could complicate this code further to address the specific problem I opened with, but I really think that, from a control flow perspective, the original code I submitted is simpler and better. If the duplication is that much of a problem, I can pull out the few lines of page turn code into local functions.

About page fit... honestly, I was hoping to punt refining page fit in the new modes. The current page fit algorithm seems to take into account horizontal margin but not vertical margin. In horizontal scrolling mode, it would presumably be straightforward to tweak that algorithm to do the opposite. But I'm not sure what should be done with spread locks on; should page fit fit to a single page or to a spread? If a single page, can it ignore both horizontal and vertical margins, since there'll likely be a scroll bar in both orientations? I wanted to defer this debate because the implementation here is pretty usable even if page fit isn't perfect, and the default viewing mode doesn't regress.

Reply:
The main objective is not getting rid of code duplication, but trying to achieve some consistency in what the shortcuts do regardless of whether vertical or horizontal scrolling is used.

Currently, in page-fit mode, backspace / Enter / space bar can be used to navigate to the previous/next page. In your current PR with horizontal scrolling, the presence of the horizontal scrollbar prevents these shortcuts from working (this is the `.pdfViewer.isHorizontalScrollbarEnabled` case, that I failed to address again).

I agree with not blocking on page spread due to the complexity.
However, for horizontal scrolling without spreading, I believe that page fit should work as expected, i.e. the height should fit. Without it, the down/up arrow keys won't result in navigations between pages, which reduces the usability of horizontal scroll mode + page fit.

Reply:
Okay, I see. I am on board with making backspace, enter, and space bar all continue to turn pages when in page fit mode regardless of scroll mode, so I'll do what it takes to make that work.